lib/linguist/heuristics.rb

```diff
@@ -73,6 +73,14 @@ def call(data)
     # Common heuristics
     ObjectiveCRegex = /^\s*(@(interface|class|protocol|property|end|synchronised|selector|implementation)\b|#import\s+.+\.h[">])/
 
+    disambiguate ".as" do |data|
+      if /^\s*(package\s+[a-z0-9_\.]+|import\s+[a-zA-Z0-9_\.]+;|class\s+[A-Za-z0-9_]+\s+extends\s+[A-Za-z0-9_]+)/.match(data)
```

Comment:
Caveat pre-emptor: I know nothing about either AngelScript or ActionScript....

How confident are you that this will match most of the ActionScript files and won't match any AngelScript?

Is there a similar AngelScript-specific regex we could use to be even more precise with the "else"?  It's not essential, just trying to be extra safe.

Reply:
I don't know a lot about Actionscript myself. I actually wrote the samples myself after doing a bit of research on the language. During that I've found that you need to have a package (which is the first match here), in which any classes will be contained. (This seems like a requirement, though [this Adobe packages article](http://www.adobe.com/devnet/actionscript/learning/as3-fundamentals/packages.html) doesn't seem to mention it as a requirement.) Either way, this is exclusive to Actionscript.

The import statement in Actionscript is used like:
```
import com.example.ClassName;
```

Angelscript *also* has an import keyword, which is used like:
```
import void MyFunction(int a, int b) from "Another module";
```

Angelscript however only lets you import functions, so that shouldn't clash with the regex used in the heuristic.

Class inheritance in Actionscript uses the `extends` keyword which is not a valid keyword in Angelscript. To inherit in Angelscript, you simple use `:`.

I can come up with a few Angelscript-specific matches, but those are not required to be used at all, such as: `[A-Za-z0-9_::]+@` for reference types, `@[A-Za-z0-9_::]\s*=` for setting references instead of values, or also even using the `auto` keyword, perhaps with a match like `^\s*auto\s+`, although it's still possible that someone names their Actionscript class `auto`, or if that appears in a block comment.

There's also the `is` operator for reference comparing, so a match like `^\s*if\s*\(.*\s+is\s+` could maybe even work. (Only `\s+is\s+` probably won't, since that can appear in comments/strings.)

Would love to know your thoughts on this, I'm unsure what's normally the most accepted way of detecting these.

Reply:
Wow! This is just the substantiation I was looking for. Thanks.

Reply:
Good to hear! If you do want me to do any additional changes to the heuristic, let me know.

Reply:
I used to be highly fluent in ActionScript back in the days when Flash were relevant, and I can honestly tell you I've never seen a line like `import void MyFunction(int a, int b) from "Another module";` appear in any conventional AS3 files. It's pretty much full-on Java-esque with their dot-delimited import style:

~~~actionscript
import com.Java.*
import com.Gosling.Ryan.Cafebabe;
import com.oracle.CorporateTakeover;
~~~

Reply:
I think you've misread, that line is an Angelscript line, not Actionscript. The language names are similar I admit.. :D

Reply:
So @Alhadis, what do you think, should I change anything in the heuristic, or is this one fine?

Reply:
Ah sorry, I missed this. No, everything's fine. I was simply lending assurance that this was the correct approach, since you'd both mentioned you didn't know ActionScript.

> Also, does the automated testing test the heuristics against samples?

It does, but I believe they need to be included in [`test/test_heuristics.rb`](https://github.com/github/linguist/blob/master/test/test_heuristics.rb).

Reply:
I've added the test for the heuristic and it looks like it works.